### PR TITLE
Fix referencing video files from any path

### LINKED ISSUE #33: Local videos not getting copied as images
I have a nested hierarchy and local videos, that do not work as they are not getting copied as images are. Let me know if you need small env to reproduce.

### PR DESCRIPTION
**Pull request by QuLogic**
This makes it no longer necessary to place video files in `_static`, but instead in any location.

I spent a while going through Shinx's implementation for images, and this copies what it does:

1. File paths are normalized with `Environment.relfn2path`, and also a dependency is added for the document in the `Environment`. This allows paths relative to any document.
2. A post transform adds the files from the `Environment` to a tracker on the `Builder`. This causes the HTML5 builder to copy the file to the `_images` directory (and a unique name is guaranteed by the tracker on the `Environment`).
3. The URI is re-written to be relative to the builder image path (only when using HTML output.)

Fixes #33
Because Sphinx then tries to copy the file, it raises a warning if not found, so I think this also fixes (partially) #14. However, if you expected the file to be in the `_static` directory, it might not warn. I did not check that.

### CODE REVIEW

**Comment by severin-lemaignan** [comment]
I've tested the patch and it works brilliantly for me (also with 2 video sources). Thanks!

File: sphinxcontrib/video.py [hunk 1]
```diff
@@ -148,6 +153,27 @@ def run(self) -> List[video_node]:
...
 
+class VideoPostTransform(SphinxPostTransform):
+    """Ensure video files are copied to build directory.
+
+    This copies what is done for images in the post_process_image method of
+    sphinx.builders.Builder, except as a Transform since we can't hook into that method
+    directly.
+    """
+
+    default_priority = 200
```
Comment (12rambau) [inline sphinxcontrib/video.py:164]: how did you choose the value of the default priority ? Are you positive it cannot clash with other vanilla SphinxPostTransform object ?

File: sphinxcontrib/video.py [hunk 2]
```diff
@@ -148,6 +153,27 @@ def run(self) -> List[video_node]:
...
+
+    def run(self):
+        """Add video files to Builder's image tracking.
+
+        Doing so ensures that the builder copies the files to the output directory.
+        """
+        for node in self.document.findall(video_node):
```
Comment (12rambau) [inline sphinxcontrib/video.py:171]: `findall` is not available if `docutils<0.18.1` Can you add support for older versions using the same trick as the one we use in PST ? 
```python
def traverse_or_findall(node: Node, condition: str, **kwargs) -> Iterator[Node]:
    """Triage node.traverse (docutils <0.18.1) vs node.findall.

    TODO: This check can be removed when the minimum supported docutils version
    for numpydoc is docutils>=0.18.1.
    """
    return (
        node.findall(condition, **kwargs)
        if hasattr(node, "findall")
        else node.traverse(condition, **kwargs)
    )
```

**Review by 12rambau — CHANGES_REQUESTED** [review]
I really like this update, aligning on how `image` directive is working is a very good thing. 
Could you update the documentation to reflect this change there as well ? 

I also see that mypy is unhappy with typing checks can you have a look to the latest build ?

File: sphinxcontrib/video.py (hunk 1 above)
Comment (QuLogic) [reply sphinxcontrib/video.py:164]: To be honest, I don't quite remember. Looking at [docutils](https://docutils.sourceforge.io/docs/ref/transforms.html#transform-priority-range-categories), a 200 puts it in the "very early" stage.

It is a lower value / higher priority than most of Sphinx's builtin transforms, except for [`ReferencesResolver`](https://github.com/sphinx-doc/sphinx/blob/ae51974e217cc7590e1b9c68f18af3e043027c6d/sphinx/transforms/post_transforms/__init__.py#L61), [`OnlyNodeTransform`](https://github.com/sphinx-doc/sphinx/blob/ae51974e217cc7590e1b9c68f18af3e043027c6d/sphinx/transforms/post_transforms/__init__.py#L226), [`ImageDownloader`](https://github.com/sphinx-doc/sphinx/blob/ae51974e217cc7590e1b9c68f18af3e043027c6d/sphinx/transforms/post_transforms/images.py#L47), [`CitationReferenceTransform`](https://github.com/sphinx-doc/sphinx/blob/ae51974e217cc7590e1b9c68f18af3e043027c6d/sphinx/builders/latex/transforms.py#L535), [`MathReferenceTransform`](https://github.com/sphinx-doc/sphinx/blob/ae51974e217cc7590e1b9c68f18af3e043027c6d/sphinx/builders/latex/transforms.py#L556) and [`ViewCodeAnchorTransform`](https://github.com/sphinx-doc/sphinx/blob/ae51974e217cc7590e1b9c68f18af3e043027c6d/sphinx/ext/viewcode.py#L189). It is equal priority to [`SigElementTransformFallback`](https://github.com/sphinx-doc/sphinx/blob/ae51974e217cc7590e1b9c68f18af3e043027c6d/sphinx/transforms/post_transforms/__init__.py#L239), and [`PropagateDescDomain`](https://github.com/sphinx-doc/sphinx/blob/ae51974e217cc7590e1b9c68f18af3e043027c6d/sphinx/transforms/post_transforms/__init__.py#L282).

I think it makes sense to be higher value / lower priority to `OnlyNodeTransform`, and possibly `ImageTransform`. I don't _think_ the rest matter?

File: sphinxcontrib/video.py (hunk 2 above)
Comment (QuLogic) [reply sphinxcontrib/video.py]: OK, done.

**Comment by 12rambau** [comment]
I'll merge it with master and create a release candidate to let people play with it. As it's changing how images are referenced it's important to give people a heads up

**Comment by QuLogic** [comment]
The mypy issue seems unrelated; I can reproduce it on `master`. I see previous workflows passed, so maybe something in Sphinx changed recently?

**Comment by QuLogic** [comment]
See #37.

**Comment by Calinou** [comment]
I see the version was bumped to 0.2.1rc0 in https://github.com/sphinx-contrib/video/commit/098b467ec9b376df9c764e0a37c543926e7452ee. Could this be published on pypi (perhaps as 0.2.1) so that it can be referenced in Sphinx repositories? This would unblock me for integrating videos on godot-docs. Thanks in advance :slightly_smiling_face:

**Comment by QuLogic** [comment]
It is on PyPI: https://pypi.org/project/sphinxcontrib-video/0.2.1rc0/ (you'd need to pass `--pre` to `pip install` to see it, or explicitly ask for that version.)

**Pull request merged**